File: software/examples/delphi/ExampleThreshold.pas
program ExampleThreshold;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletHumidity;

type
  TExample = class
  private
    ipcon: TIPConnection;
    h: TBrickletHumidity;
  public
    procedure ReachedCB(sender: TBrickletHumidity; const humidity: word);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = '9TU'; { Change to your UID }

var
  e: TExample;

{ Callback for humidity outside of 30 to 60 %RH }
procedure TExample.ReachedCB(sender: TBrickletHumidity; const humidity: word);
begin
  if (humidity < 30*10) then begin
    WriteLn(Format('Humidity too low: %f %%RH', [humidity/10.0]));
  end
  else begin
    WriteLn(Format('Humidity too high: %f %%RH', [humidity/10.0]));
  end;
  WriteLn('Recommended humiditiy for human comfort is 30 to 60 %RH.');
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  h := TBrickletHumidity.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Get threshold callbacks with a debounce time of 10 seconds (10000ms) }
  h.SetDebouncePeriod(10000);

  { Register threshold reached callback to procedure ReachedCB }
  h.OnHumidityReached := {$ifdef FPC}@{$endif}ReachedCB;

  { Configure threshold for "outside of 30 to 60 %RH" (unit is %RH/10) }
  h.SetHumidityCallbackThreshold('o', 30*10, 60*10);

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy;
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/javascript/ExampleThreshold.js
var IPConnection = require('Tinkerforge/IPConnection');
var BrickletHumidity = require('Tinkerforge/BrickletHumidity');

var HOST = 'localhost';
var PORT = 4223;
var UID = '7bA'; //Change to your UID

var ipcon = new IPConnection(); //Create IP connection
var h = new BrickletHumidity(UID, ipcon); //Create device object

ipcon.connect(HOST, PORT,
    function(error) {
        if(error === IPConnection.ERROR_ALREADY_CONNECTED) {
            console.log('Error: Already connected');        
        }
    }
);//Connect to brickd

//Don't use device before ipcon is connected
ipcon.on(IPConnection.CALLBACK_CONNECTED,
    function(connectReason) {
        // Set threshold callbacks with a debounce time of 10 seconds (10000ms)
        h.setDebouncePeriod(10000); 
        // Configure threshold for "outside of 30 to 60 %RH" (unit is %RH/10)
        h.setHumidityCallbackThreshold('o', 30*10, 60*10);    
    }
);

// Register threshold reached callback
h.on(BrickletHumidity.CALLBACK_HUMIDITY_REACHED,
    // Callback for humidity outside of 30 to 60 %RH
    function(humidity) {
        if(humidity < 30*10) {
            console.log('Humidity too low: '+humidity/10+' %RH');
        }
        if(humidity > 60*10) {
            console.log('Humidity too high: '+humidity/10+' %RH');
        }
    }
);

console.log("Press any key to exit ...");
process.stdin.on('data', function(data) {
	    ipcon.disconnect(
            function(error) {
                if(error === IPConnection.ERROR_NOT_CONNECTED) {
                    console.log('Error: Not connected');        
                }
            }
        );
process.exit(0);
});

